### DISCUSSION TITLE
Fixing application issues (hidden or not launching)

### DISCUSSION THREAD

**Discussion by friday** [Ulauncher/Ulauncher#1348] [Troubleshooting]
Ulauncher uses GTK's APIs to handle applications. If your applications doesn't show up in Ulauncher, this is not an issue with Ulauncher. It is likely because the desktop entry file is in the wrong path, doesn't comply with the freedesktop specifications, has an invalid path to the executable, or that it's intentionally hidden with `NoDisplay`, `OnlyShowIn`, `NotShowIn` or `Hidden`.

By default (unless you have specified otherwise in your `XDG_DATA_DIRS` or `XDG_DATA_HOME` variable) the paths the application indexer is searching are `/usr/share/applications/` and `~/.local/share/applications`. The former is where system packages tend to install desktop entries. These are rarely invalid/broken, but the home dir path overrides the ones in the system path. So if you have a file with the same name in both these directories the second one will be used. If that one is invalid, it might not show up even if the other one is valid.

To find the issue, first find the desktop entry file(s) for the app(s) that are not showing. If you can't find it in the paths previously mentioned, this is likely the cause of the issue. It's possible to configure your system differently so the locations would be elsewhere, but we assume any users doing so would be able to "translate" these instructions for their paths.

If you do find the file, first check that the command to launch the app as specified in the desktop entry's <code>Exec</code> line corresponds to an actual existing file and that it has the execute permission.

If this doesn't fix the issue, make sure you have [`desktop-file-utils`](https://www.freedesktop.org/wiki/Software/desktop-file-utils/). Then you can validate the file(s).

This command should show all validation issues for all your local desktop entries:

```sh
desktop-file-validate  ~/.local/share/applications/**/*.desktop
```

Fix these issues manually to fix your applications.

**Comment by beobcode** [comment]
I had an issue with the OnlyShowIn key set. I removed the key, then the application showed up in my search results.

**Reply by friday** [reply]
That's how OnlyShowIn should work: https://specifications.freedesktop.org/menu-spec/latest/ar01s03.html, but I guess it's a good addition to this troubleshooting page :slightly_smiling_face:

**Reply by beobcode** [reply]
I think it will be a good idea. I followed the instruction and the file was valid. It took a bit of thinking about looking at different files until I noticed it. 
I was also thinking about NoDisplay and Hidden, but I'm not sure how that affects Ulauncher. I have not done the test myself yet.

**Reply by friday** [reply]
Updated :+1: 

This happens because you are trying to use an app that came bundled with a specific desktop environment, but you are not currently using that desktop environment. For example XFCE's system settings should not run in Gnome. And if you have 3-5 DEs installed on your system and search for apps with generic names like "File manager" "Text editor" etc it would be annoying if it showed 5 apps with the same name and icon and you don't know which one is for your DE.

To edit the desktop entry, make sure to copy it to ~/.local/share/applications first. Otherwise it will be overwritten if you update the app.

**Reply by beobcode** [reply]
Oh. Thanks for your help. I worked out the issues I was experiencing now.  I use Cinnamon, and I was trying to add the gnome-terminal to my search results. This had the key `OnlyShowIn=GNOME;Unity;` I just removed it and it worked.  But with the information you provided, I was able to work out that I can just add "X-Cinnamon" so the key would be `OnlyShowIn=GNOME;Unity;X-Cinnamon;`
To be honest, what confused me was the error Ulauncher gave me (mate-about.desktop was a test file)
`2024-07-19 09:51:51,843 | WARNING | ulauncher.search.apps.app_watcher: _add_file_sync() | Cannot add /home/user/.local/share/applications/mate-about.desktop to DB -> /home/user/.local/share/applications/mate-about.desktop`

**Reply by friday** [reply]
That part no longer exists in the v6 branch though. Not that saying nothing is less confusing.

**Comment by Lorak-mmk** [comment]
I have an issue with `Kate` (text editor from KDE) not showing up (ULauncher on Fedora 41).
This seems to mostly be Kate's fault, on each launch it produces desktop in `~/.local/share/applications` file that is a bit incorrect:
```
bat /home/lorak/.local/share/applications/org.kde.kate.desktop
───────┬────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
       │ File: /home/lorak/.local/share/applications/org.kde.kate.desktop
───────┼────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
   1   │ [Desktop Action Session 7a1920d61156abc05a60135aefe8bc67]
   2   │ Exec=kate -n -s Default
   3   │ Name=Default
   4   │ 
   5   │ [Desktop Entry]
   6   │ Actions=Session 7a1920d61156abc05a60135aefe8bc67;
   7   │ Categories=Qt;KDE;Utility;TextEditor;Development;
   8   │ Comment=KDE Advanced Text Editor
   9   │ Exec=kate -b %U
  10   │ GenericName=Advanced Text Editor
  11   │ Icon=kate
  12   │ InitialPreference=9
  13   │ Keywords=text;txt;editor;programming;programmer;development;developer;code;
  14   │ MimeType=text/plain;inode/directory;
  15   │ Name=Kate
  16   │ StartupNotify=true
  17   │ StartupWMClass=kate
  18   │ Type=Application
  19   │ X-DocPath=kate/index.html
  20   │ X-KDE-HasTempFileOption=true
───────┴────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
```

```
desktop-file-validate /home/lorak/.local/share/applications/org.kde.kate.desktop
/home/lorak/.local/share/applications/org.kde.kate.desktop: error: first group is not "Desktop Entry"
/home/lorak/.local/share/applications/org.kde.kate.desktop: error: file contains group "Desktop Action Session 7a1920d61156abc05a60135aefe8bc67", which has an invalid action identifier, only alphanumeric characters and '-' are allowed
/home/lorak/.local/share/applications/org.kde.kate.desktop: error: value "Session 7a1920d61156abc05a60135aefe8bc67;" for key "Actions" in group "Desktop Entry" contains invalid action identifier "Session 7a1920d61156abc05a60135aefe8bc67", only alphanumeric characters and '-' are allowed
/home/lorak/.local/share/applications/org.kde.kate.desktop: hint: value "Qt;KDE;Utility;TextEditor;Development;" for key "Categories" in group "Desktop Entry" contains more than one main category; application might appear more than once in the application menu
```


However, it would be nice of Ulauncher if in case of such wrong desktop file it used the other one - `/usr/share/applications/org.kde.kate.desktop` is correct and present, yet ignored, so Kate is not showing in my ulauncher at all :(

I also don't see the option to ignore specific files, only directories, so I can't really work around this issue.

**Comment by OWL4C** [comment]
I am just posting this here so that other people that search for it may find it more easily, but after updating my Zen Browser AppImage (integrated via Gear Lever) via the KDE Switcher it moved the `[Desktop Entry]` Section behind the `[Desktop Action <>]` sections, as well as adding `TryExec=` lines, and ulauncher couldn't find it anymore.
Moving the Desktop Entry back at the top and removing all `TryExec` lines fixed it.
